lib/OpenQA/WebAPI.pm

```diff
@@ -553,5 +571,110 @@ sub run {
     Mojolicious::Commands->start_app('OpenQA::WebAPI');
 }
 
+sub get_pod_from_controllers {
+    # Object to get API descriptions from POD
+    my $tree = Pod::Tree->new or die "cannot create object: $!\n";
+    my %controllers;
+    my $OPENQA_CODEBASE = '/usr/share/openqa';
+    my $ctrlrpath = path($OPENQA_CODEBASE)->child('lib', 'OpenQA', 'WebAPI', 'Controller', 'API', 'V1');
+
+    # Review all routes to get controllers, and from there get the .pm filename to parse for POD
+    foreach my $api_rt (@_) {
+        next if (ref($api_rt) ne 'Mojolicious::Routes::Route');
+        foreach my $rt (@{$api_rt->children}) {
+            next unless ($rt->to->{controller});
+            my $filename = ucfirst($rt->to->{controller});
+            if ($filename =~ /_/) {
+                $filename = join('', map(ucfirst, split(/_/, $filename)));
+            }
+            $filename .= '.pm';
+            $controllers{$rt->to->{controller}} = $filename;
+        }
+    }
+
+    # Parse API controller files for POD
+    foreach my $ctrl (keys %controllers) {
+        $tree->load_file($ctrlrpath->child($controllers{$ctrl})->to_string);
+        $tree->get_root->set_filename($ctrl);
+        if ($tree->loaded() and $tree->has_pod()) {
+            $tree->walk(\&_itemize);
+        }
+    }
+}
+
+sub set_api_desc {
+    my $api_description = shift;
+    my $api_route       = shift;
+
+    if (ref($api_description) ne 'HASH') {
+        log_warning("set_api_desc: expected HASH ref for api_descriptions. Got: " . ref($api_description));
+        return;
+    }
+
+    if (ref($api_route) ne 'Mojolicious::Routes::Route') {
+        log_warning("set_api_desc: expected Mojolicious::Routes::Route for api_routes. Got: " . ref($api_route));
+        return;
+    }
+
+    foreach my $r (@{$api_route->children}) {
+        next unless ($r->to->{controller} and $r->to->{action});
+        my $key = $r->to->{controller} . '#' . $r->to->{action};
+        $api_description->{$r->name} = $methods_description{$key} if (defined $methods_description{$key});
+    }
+}
+
+sub _itemize {
+    if (ref($_[0]) ne 'Pod::Tree::Node') {
+        log_warning("_itemize() expected Pod::Tree::Node arg. Got " . ref($_[0]));
+        return 0; # Stop walking the tree
+    }
+    my $methodname = '';
+    my $desc       = '';
+    my $controller = $_[0]->get_filename;
+    if ($_[0]->is_item()) {
+        $methodname = _get_pod_text($_[0]);
+        $methodname =~ s/\s+//g;
+        $methodname =~ s/\(\)//;
+        my $siblings = $_[0]->get_siblings();
+        foreach my $i (@$siblings) {
+            unless ($desc) { # Only take first paragraph for the description
+                $desc = $i->get_text() if $i->is_text();
+                $desc = _get_pod_text($i) if $i->is_ordinary();
+            }
+        }
+        $methods_description{$controller . '#' . $methodname} = $desc;
```

Comment:
I'd suggest to build the srting first, and then use it to avoid perl critic complains: https://travis-ci.org/os-autoinst/openQA/jobs/327749143#L652

Reply:
Cool. Will do that.